.gitlab-ci.yml

```diff
@@ -728,44 +729,81 @@ publish-rustdoc:
   stage:                           publish
   <<:                              *kubernetes-env
   <<:                              *vault-secrets
-  image:                           paritytech/tools:latest
+  image:                           node:16
   variables:
     GIT_DEPTH:                     100
+    # --- Following variables are for rustdocs deployment ---
+    # Space separated values of branches and tags to generate rustdocs
+    RUSTDOCS_DEPLOY_REFS:   "master monthly-2021-09+1 monthly-2021-08 v3.0.0"
+    # Location of the docs index template
+    INDEX_TPL:              ".maintain/docs-index-tpl.ejs"
+    # Where the `/latest` symbolic link links to. One of the $RUSTDOCS_DEPLOY_REFS value.
+    LATEST:                 "monthly-2021-09+1"
+
   rules:
     - if: $CI_PIPELINE_SOURCE == "pipeline"
       when: never
-    - if: $CI_PIPELINE_SOURCE == "web" && $CI_COMMIT_REF_NAME == "master"
+    # Manually add branches based on $RUSTDOCS_DEPLOY_REFS
+    # Maybe there is a smarter way?
     - if: $CI_COMMIT_REF_NAME == "master"
+    - if: $CI_COMMIT_REF_NAME == "monthly-2021-09+1"
+    - if: $CI_COMMIT_REF_NAME == "monthly-2021-08"
+    - if: $CI_COMMIT_REF_NAME == "v3.0.0"
+
   # `needs:` can be removed after CI image gets nonroot. In this case `needs:` stops other
   # artifacts from being dowloaded by this job.
   needs:
     - job:                         build-rustdoc
       artifacts:                   true
-  script:
-    - rm -rf /tmp/*
+  script: |
+    rm -rf /tmp/*
+
     # Set git config
-    - rm -rf .git/config
-    - git config user.email "devops-team@parity.io"
-    - git config user.name "${GITHUB_USER}"
-    - git config remote.origin.url "https://${GITHUB_TOKEN}@github.com/paritytech/substrate.git"
-    - git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
-    - git fetch origin gh-pages
+    rm -rf .git/config
+    git config user.email "devops-team@parity.io"
+    git config user.name "${GITHUB_USER}"
+    git config remote.origin.url "https://${GITHUB_TOKEN}@github.com/paritytech/${CI_PROJECT_NAME}.git"
+    git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
+    git fetch origin gh-pages
+
+    # Install `ejs` and generate index.html based on RUSTDOCS_DEPLOY_REFS
+    yarn global add ejs
```

Comment:
is there any way to not install something on every CI run?

Reply:
Probably we can do caching here, but not sure where to cache.